docs/man/qmicli.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.46.4.
.TH QMICLI "1" "February 2015" "qmicli " "User Commands"
.SH NAME
qmicli \- Control QMI devices
.SH DESCRIPTION
.SS "Usage:"
.IP
lt\-qmicli [OPTION...] \- Control QMI devices
.SS "Help Options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Show help options
.TP
\fB\-\-help\-all\fR
Show all help options
.TP
\fB\-\-help\-dms\fR
Show Device Management Service options
.TP
\fB\-\-help\-nas\fR
Show Network Access Service options
.TP
\fB\-\-help\-wds\fR
Show Wireless Data Service options
.TP
\fB\-\-help\-pbm\fR
Show Phonebook Management options
.TP
\fB\-\-help\-uim\fR
Show User Identity Module options
.TP
\fB\-\-help\-wda\fR
Show Wireless Data Administrative options
.PP
DMS options
.TP
\fB\-\-dms\-get\-ids\fR
Get IDs
.TP
\fB\-\-dms\-get\-capabilities\fR
Get capabilities
.TP
\fB\-\-dms\-get\-manufacturer\fR
Get manufacturer
.TP
\fB\-\-dms\-get\-model\fR
Get model
.TP
\fB\-\-dms\-get\-revision\fR
Get revision
.TP
\fB\-\-dms\-get\-msisdn\fR
Get MSISDN
.TP
\fB\-\-dms\-get\-power\-state\fR
Get power state
.TP
\fB\-\-dms\-uim\-set\-pin\-protection\fR=\fI\,[\/\fR(PIN|PIN2),(disable|enable),(current PIN)]
Set PIN protection in the UIM
.TP
\fB\-\-dms\-uim\-verify\-pin\fR=\fI\,[\/\fR(PIN|PIN2),(current PIN)]
Verify PIN
.TP
\fB\-\-dms\-uim\-unblock\-pin\fR=\fI\,[\/\fR(PIN|PIN2),(PUK),(new PIN)]
Unblock PIN
.TP
\fB\-\-dms\-uim\-change\-pin\fR=\fI\,[\/\fR(PIN|PIN2),(old PIN),(new PIN)]
Change PIN
.TP
\fB\-\-dms\-uim\-get\-pin\-status\fR
Get PIN status
.TP
\fB\-\-dms\-uim\-get\-iccid\fR
Get ICCID
.TP
\fB\-\-dms\-uim\-get\-imsi\fR
Get IMSI
.TP
\fB\-\-dms\-uim\-get\-state\fR
Get UIM State
.TP
\fB\-\-dms\-uim\-get\-ck\-status\fR=\fI\,[\/\fR(pn|pu|pp|pc|pf)]
Get CK Status
.TP
\fB\-\-dms\-uim\-set\-ck\-protection\fR=\fI\,[\/\fR(pn|pu|pp|pc|pf),(disable),(key)]
Disable CK protection
.TP
\fB\-\-dms\-uim\-unblock\-ck\fR=\fI\,[\/\fR(pn|pu|pp|pc|pf),(key)]
Unblock CK
.TP
\fB\-\-dms\-get\-hardware\-revision\fR
Get the HW revision
.TP
\fB\-\-dms\-get\-operating\-mode\fR
Get the device operating mode
.TP
\fB\-\-dms\-set\-operating\-mode\fR=\fI\,[\/\fR(Operating mode)]
Set the device operating mode
.TP
\fB\-\-dms\-get\-time\fR
Get the device time
.TP
\fB\-\-dms\-get\-prl\-version\fR
Get the PRL version
.TP
\fB\-\-dms\-get\-activation\-state\fR
Get the state of the service activation
.TP
\fB\-\-dms\-activate\-automatic\fR=\fI\,[Activation\/\fR Code]
Request automatic service activation
.TP
\fB\-\-dms\-activate\-manual\fR=\fI\,[SPC\/\fR,SID,MDN,MIN]
Request manual service activation
.TP
\fB\-\-dms\-get\-user\-lock\-state\fR
Get the state of the user lock
.TP
\fB\-\-dms\-set\-user\-lock\-state\fR=\fI\,[\/\fR(disable|enable),(current lock code)]
Set the state of the user lock
.TP
\fB\-\-dms\-set\-user\-lock\-code\fR=\fI\,[\/\fR(old lock code),(new lock code)]
Change the user lock code
.TP
\fB\-\-dms\-read\-user\-data\fR
Read user data
.TP
\fB\-\-dms\-write\-user\-data\fR=\fI\,[\/\fR(User data)]
Write user data
.TP
\fB\-\-dms\-read\-eri\-file\fR
Read ERI file
.TP
\fB\-\-dms\-restore\-factory\-defaults\fR=\fI\,[\/\fR(Service Programming Code)]
Restore factory defaults
.TP
\fB\-\-dms\-validate\-service\-programming\-code\fR=\fI\,[\/\fR(Service Programming Code)]
Validate the Service Programming Code
.TP
\fB\-\-dms\-get\-band\-capabilities\fR
Get band capabilities
.TP
\fB\-\-dms\-get\-factory\-sku\fR
Get factory stock keeping unit
.TP
\fB\-\-dms\-list\-stored\-images\fR
List stored images
.TP
\fB\-\-dms\-select\-stored\-image\fR=\fI\,[modem\/\fR#,pri#] where # is the index
Select stored image
.TP
\fB\-\-dms\-delete\-stored\-image\fR=\fI\,[modem\/\fR#|pri#] where # is the index
Delete stored image
.TP
\fB\-\-dms\-set\-fcc\-authentication\fR
Set FCC authentication
.TP
\fB\-\-dms\-reset\fR
Reset the service state
.TP
\fB\-\-dms\-noop\fR
Just allocate or release a DMS client. Use with `\-\-client\-no\-release\-cid' and/or `\-\-client\-cid'
.PP
NAS options
.TP
\fB\-\-nas\-get\-signal\-strength\fR
Get signal strength
.TP
\fB\-\-nas\-get\-signal\-info\fR
Get signal info
.TP
\fB\-\-nas\-get\-tx\-rx\-info\fR=\fI\,[\/\fR(Radio Interface)]
Get TX/RX info
.TP
\fB\-\-nas\-get\-home\-network\fR
Get home network
.TP
\fB\-\-nas\-get\-serving\-system\fR
Get serving system
.TP
\fB\-\-nas\-get\-system\-info\fR
Get system info
.TP
\fB\-\-nas\-get\-technology\-preference\fR
Get technology preference
.TP
\fB\-\-nas\-get\-system\-selection\-preference\fR
Get system selection preference
.TP
\fB\-\-nas\-set\-system\-selection\-preference\fR=\fI\,[cdma\-1x\/\fR|cdma\-1xevdo|gsm|umts|lte|td\-scdma]
Set system selection preference
.TP
\fB\-\-nas\-network\-scan\fR
Scan networks
.TP
\fB\-\-nas\-get\-cell\-location\-info\fR
Get Cell Location Info
.TP
\fB\-\-nas\-reset\fR
Reset the service state
.TP
\fB\-\-nas\-noop\fR
Just allocate or release a NAS client. Use with `\-\-client\-no\-release\-cid' and/or `\-\-client\-cid'
.PP
WDS options
.TP
\fB\-\-wds\-start\-network\fR=\fI\,[\/\fR(APN),(PAP|CHAP|BOTH),(Username),(Password)]
Start network (Authentication, Username and Password are optional)
.TP
\fB\-\-wds\-follow\-network\fR
Follow the network status until disconnected. Use with `\-\-wds\-start\-network'
.TP
\fB\-\-wds\-stop\-network\fR=\fI\,[Packet\/\fR data handle]
Stop network
.TP
\fB\-\-wds\-get\-packet\-service\-status\fR
Get packet service status
.TP
\fB\-\-wds\-get\-packet\-statistics\fR
Get packet statistics
.TP
\fB\-\-wds\-get\-data\-bearer\-technology\fR
Get data bearer technology
.TP
\fB\-\-wds\-get\-current\-data\-bearer\-technology\fR
Get current data bearer technology
.TP
\fB\-\-wds\-get\-profile\-list\fR=\fI\,[3gpp\/\fR|3gpp2]
Get profile list
.TP
\fB\-\-wds\-get\-default\-settings\fR=\fI\,[3gpp\/\fR|3gpp2]
Get default settings
.TP
\fB\-\-wds\-reset\fR
Reset the service state
.TP
\fB\-\-wds\-noop\fR
Just allocate or release a WDS client. Use with `\-\-client\-no\-release\-cid' and/or `\-\-client\-cid'
.PP
PBM options
.TP
\fB\-\-pbm\-get\-all\-capabilities\fR
Get all phonebook capabilities
.TP
\fB\-\-pbm\-noop\fR
Just allocate or release a PBM client. Use with `\-\-client\-no\-release\-cid' and/or `\-\-client\-cid'
.PP
UIM options
.TP
\fB\-\-uim\-read\-transparent\fR=\fI\,[0xNNNN\/\fR,0xNNNN,...]
Read a transparent file given the file path
.TP
\fB\-\-uim\-get\-file\-attributes\fR=\fI\,[0xNNNN\/\fR,0xNNNN,...]
Get the attributes of a given file
.TP
\fB\-\-uim\-get\-card\-status\fR
Get card status
.TP
\fB\-\-uim\-reset\fR
Reset the service state
.TP
\fB\-\-uim\-noop\fR
Just allocate or release a UIM client. Use with `\-\-client\-no\-release\-cid' and/or `\-\-client\-cid'
.PP
WDA options
.TP
\fB\-\-wda\-set\-data\-format\fR=\fI\,[raw\-ip\/\fR|802\-3]
Set data format
.TP
\fB\-\-wda\-get\-data\-format\fR
Get data format
.TP
\fB\-\-wda\-noop\fR
Just allocate or release a WDA client. Use with `\-\-client\-no\-release\-cid' and/or `\-\-client\-cid'
.SS "Application Options:"
.TP
\fB\-d\fR, \fB\-\-device\fR=\fI\,[PATH]\/\fR
Specify device path
.TP
\fB\-\-get\-service\-version\-info\fR
Get service version info
.TP
\fB\-\-device\-set\-instance\-id\fR=\fI\,[Instance\/\fR ID]
Set instance ID
.TP
\fB\-\-device\-open\-version\-info\fR
Run version info check when opening device
.TP
\fB\-\-device\-open\-sync\fR
Run sync operation when opening device
.TP
\fB\-p\fR, \fB\-\-device\-open\-proxy\fR
Request to use the 'qmi\-proxy' proxy
.TP
\fB\-\-device\-open\-net\fR=\fI\,[net\-802\-3\/\fR|net\-raw\-ip|net\-qos\-header|net\-no\-qos\-header]
Open device with specific link protocol and QoS flags
.TP
\fB\-\-client\-cid\fR=\fI\,[CID]\/\fR
Use the given CID, don't allocate a new one
.TP
\fB\-\-client\-no\-release\-cid\fR
Do not release the CID when exiting
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Run action with verbose logs, including the debug ones
.TP
\fB\-\-silent\fR
Run action with no logs; not even the error/warning ones
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version
.PP
qmicli 1.12.6
Copyright \(co 2015 Aleksander Morgado
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl\-2.0.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
The full documentation for
.B qmicli
is maintained as a Texinfo manual.  If the
.B info
and
.B qmicli
programs are properly installed at your site, the command
.IP
.B info qmicli
.PP
should give you access to the complete manual.


debian/changelog
libqmi (1.12.6-1) unstable; urgency=medium

  * Upload to unstable.
  * New upstream release.
  * Add myself to Uploaders.

 -- Michael Biebl <biebl@debian.org>  Thu, 07 May 2015 00:51:05 +0200

libqmi (1.12.4-2) experimental; urgency=medium

  * Team upload.
  * Mark binary packages as linux-any.

 -- Michael Biebl <biebl@debian.org>  Fri, 13 Feb 2015 21:31:59 +0100

libqmi (1.12.4-1) experimental; urgency=medium

  * Team upload.

  [ Marius B. Kotsbak ]
  * Add gbp config for experimental branch.

  [ Michael Biebl ]
  * Imported Upstream version 1.12.4.
  * Update symbols file.
  * Update Vcs-Browser URL to use cgit and https.

 -- Michael Biebl <biebl@debian.org>  Fri, 13 Feb 2015 09:43:06 +0100

libqmi (1.12.2-1) experimental; urgency=medium

  [ Marius B. Kotsbak ]
  * Imported Upstream version 1.12.2 (Closes: #772328)
  * Updated symbols file for libqmi-glib1 package for new upstream 1.12.2.
  * Update Debian standards version to 3.9.6
  * Install NEWS file which includes the relevant upstream changes.

 -- Marius B. Kotsbak <marius@kotsbak.com>  Sat, 31 Jan 2015 14:41:43 +0100

libqmi (1.10.2-2) unstable; urgency=medium

  * Team upload.

  [ Marius B. Kotsbak ]
  * Use dh_install --list-missing to show uninstalled files.
  * Drop obsolete --with-tests configure option.

  [ Michael Biebl ]
  * Install man pages via dh_install.
  * Use a strictly versioned dependency between libqmi-proxy and libqmi-glib1,
    so the two packages don't get out of sync.
  * Make any package linking against libqmi-glib1 get an explicit dependency
    on libqmi-proxy. The qmi-proxy binary is used internally by libqmi-glib1
    and the package split is an implementation detail to properly support
    multi-arch. Convince lintian that this is intentional.
  * Add Depends on libglib2.0-dev to libqmi-glib-dev as required by
    qmi-glib.pc.

 -- Michael Biebl <biebl@debian.org>  Tue, 16 Sep 2014 23:24:51 +0200

libqmi (1.10.2-1) unstable; urgency=low

  [ Marius B. Kotsbak ]
  * New upstream stable version 1.10.2
    - Fixed default internal proxy timeout for requests.
    - Added initial support for the WDA service.
    - Added support for cell location info retrieval.
    - Added support for UIM card status retrieval.
    - Added support to specify net open flags in the command line.
    - Minor bugfixes

 -- Marius B. Kotsbak <marius@kotsbak.com>  Tue, 29 Jul 2014 20:45:51 +0200

libqmi (1.8.0-2) unstable; urgency=low

  [ Marius B. Kotsbak ]
  * Change libqmi to only recommend libqmi-proxy package.
    It is possible to use the library without it (Closes: #754599).

 -- Marius B. Kotsbak <marius@kotsbak.com>  Thu, 17 Jul 2014 20:35:17 +0200

libqmi (1.8.0-1) unstable; urgency=low

  [ Marius Kotsbak ]
  * New upstream stable release 1.8.0 (Closes: #731002, #731907)
    - Support for sharing the QMI port through multiple processes with the new
      'qmi-proxy' infrastructure.
    - api: added new known service ID enumeration values
    - wds: added "Get Profile List" request/response
    - wds: added "Get Profile Settings" request/response
    - wds: added "Get Default Settings" request/response
    - cli: add --wds-get-profile-list
    - cli: add --wds-get-default-settings
    - dms: 'SID' TLV in "Activate Manual" is a guint8, not a string
    - Extended API with macros to check for the libqmi-glib version in use
    - Setup libtool versioning in libqmi-glib
    - Added initial support for the PBM (phonebook) service
    - Added initial support for the UIM service
    - Added initial support for the OMA (Open Mobile Alliance) service
    - nas: added "Get Tx Rx Info" request/response
    - wds: added "Get Packet Statistics" request/response
    - dms: added support for the 'PRL' TLV in "Activate Manual"
    - cli: add --nas-get-tx-rx-info
    - cli: add --wds-get-packet-statistics
    - cli: add --get-service-version-info
    - cli: add --uim-read-transparent
    - cli: add --uim-get-file-attributes
    - cli: fix following network status until disconnected
    - tests: use gtester to run unit tests
    - Also fixed several minor bugs and memory leaks
  * Replaced "help2man" in Debian package with the same from upstream

 -- Marius B. Kotsbak <marius@kotsbak.com>  Sat, 21 Jun 2014 08:02:00 +0200


src/qmicli/qmicli.c
/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*
 * qmicli -- Command line interface to control QMI devices
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright (C) 2012-2015 Aleksander Morgado <aleksander@aleksander.es>
 */

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <locale.h>
#include <string.h>

#include <glib.h>
#include <glib/gprintf.h>
#include <gio/gio.h>

#include <libqmi-glib.h>

#include "qmicli.h"
#include "qmicli-helpers.h"

#define PROGRAM_NAME    "qmicli"
#define PROGRAM_VERSION PACKAGE_VERSION

/* Globals */
static GMainLoop *loop;
static GCancellable *cancellable;
static QmiDevice *device;
static QmiClient *client;
static QmiService service;
static gboolean operation_status;

/* Main options */
static gchar *device_str;
static gboolean get_service_version_info_flag;
static gchar *device_set_instance_id_str;
static gboolean device_open_version_info_flag;
static gboolean device_open_sync_flag;
static gchar *device_open_net_str;
static gboolean device_open_proxy_flag;
static gboolean device_open_mbim_flag;
static gchar *client_cid_str;
static gboolean client_no_release_cid_flag;
static gboolean verbose_flag;
static gboolean silent_flag;
static gboolean version_flag;

static GOptionEntry main_entries[] = {
    { "device", 'd', 0, G_OPTION_ARG_STRING, &device_str,
      "Specify device path",
      "[PATH]"
    },
    { "get-service-version-info", 0, 0, G_OPTION_ARG_NONE, &get_service_version_info_flag,
      "Get service version info",
      NULL
    },
    { "device-set-instance-id", 0, 0, G_OPTION_ARG_STRING, &device_set_instance_id_str,
      "Set instance ID",
      "[Instance ID]"
    },
    { "device-open-version-info", 0, 0, G_OPTION_ARG_NONE, &device_open_version_info_flag,
      "Run version info check when opening device",
      NULL
    },
    { "device-open-sync", 0, 0, G_OPTION_ARG_NONE, &device_open_sync_flag,
      "Run sync operation when opening device",
      NULL
    },
    { "device-open-proxy", 'p', 0, G_OPTION_ARG_NONE, &device_open_proxy_flag,
      "Request to use the 'qmi-proxy' proxy",
      NULL
    },
    { "device-open-mbim", 0, 0, G_OPTION_ARG_NONE, &device_open_mbim_flag,
      "Open an MBIM device with EXT_QMUX support",
      NULL
    },
    { "device-open-net", 0, 0, G_OPTION_ARG_STRING, &device_open_net_str,
      "Open device with specific link protocol and QoS flags",
      "[net-802-3|net-raw-ip|net-qos-header|net-no-qos-header]"
    },
    { "client-cid", 0, 0, G_OPTION_ARG_STRING, &client_cid_str,
      "Use the given CID, don't allocate a new one",
      "[CID]"
    },
    { "client-no-release-cid", 0, 0, G_OPTION_ARG_NONE, &client_no_release_cid_flag,
      "Do not release the CID when exiting",
      NULL
    },
    { "verbose", 'v', 0, G_OPTION_ARG_NONE, &verbose_flag,
      "Run action with verbose logs, including the debug ones",
      NULL
    },
    { "silent", 0, 0, G_OPTION_ARG_NONE, &silent_flag,
      "Run action with no logs; not even the error/warning ones",
      NULL
    },
    { "version", 'V', 0, G_OPTION_ARG_NONE, &version_flag,
      "Print version",
      NULL
    },
    { NULL }
};

static void
signals_handler (int signum)
{
    if (cancellable) {
        /* Ignore consecutive requests of cancellation */
        if (!g_cancellable_is_cancelled (cancellable)) {
            g_printerr ("%s\n",
                        "cancelling the operation...\n");
            g_cancellable_cancel (cancellable);
        }
        return;
    }

    if (loop &&
        g_main_loop_is_running (loop)) {
        g_printerr ("%s\n",
                    "cancelling the main loop...\n");
        g_main_loop_quit (loop);
    }
}

static void
log_handler (const gchar *log_domain,
             GLogLevelFlags log_level,
             const gchar *message,
             gpointer user_data)
{
    const gchar *log_level_str;
    time_t now;
    gchar time_str[64];
    struct tm *local_time;
    gboolean err;

    /* Nothing to do if we're silent */
    if (silent_flag)
        return;

    now = time ((time_t *) NULL);
    local_time = localtime (&now);
    strftime (time_str, 64, "%d %b %Y, %H:%M:%S", local_time);
    err = FALSE;

    switch (log_level) {
    case G_LOG_LEVEL_WARNING:
        log_level_str = "-Warning **";
        err = TRUE;
        break;

    case G_LOG_LEVEL_CRITICAL:
    case G_LOG_FLAG_FATAL:
    case G_LOG_LEVEL_ERROR:
        log_level_str = "-Error **";
        err = TRUE;
        break;

    case G_LOG_LEVEL_DEBUG:
        log_level_str = "[Debug]";
        break;

    default:
        log_level_str = "";
        break;
    }

    if (!verbose_flag && !err)
        return;

    g_fprintf (err ? stderr : stdout,
               "[%s] %s %s\n",
               time_str,
               log_level_str,
               message);
}

static void
print_version_and_exit (void)
{
    g_print ("\n"
             PROGRAM_NAME " " PROGRAM_VERSION "\n"
             "Copyright (C) 2015 Aleksander Morgado\n"
             "License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl-2.0.html>\n"
             "This is free software: you are free to change and redistribute it.\n"
             "There is NO WARRANTY, to the extent permitted by law.\n"
             "\n");
    exit (EXIT_SUCCESS);
}

static gboolean
generic_options_enabled (void)
{
    static guint n_actions = 0;
    static gboolean checked = FALSE;

    if (checked)
        return !!n_actions;

    n_actions = (!!device_set_instance_id_str +
                 get_service_version_info_flag);

    if (n_actions > 1) {
        g_printerr ("error: too many generic actions requested\n");
        exit (EXIT_FAILURE);
    }

    checked = TRUE;
    return !!n_actions;
}

/*****************************************************************************/
/* Running asynchronously */

static void
release_client_ready (QmiDevice *dev,
                      GAsyncResult *res)
{
    GError *error = NULL;

    if (!qmi_device_release_client_finish (dev, res, &error)) {
        g_printerr ("error: couldn't release client: %s", error->message);
        g_error_free (error);
    } else
        g_debug ("Client released");

    g_main_loop_quit (loop);
}

void
qmicli_async_operation_done (gboolean reported_operation_status)
{
    QmiDeviceReleaseClientFlags flags = QMI_DEVICE_RELEASE_CLIENT_FLAGS_NONE;

    /* Keep the result of the operation */
    operation_status = reported_operation_status;

    if (cancellable) {
        g_object_unref (cancellable);
        cancellable = NULL;
    }

    /* If no client was allocated (e.g. generic action), just quit */
    if (!client) {
        g_main_loop_quit (loop);
        return;
    }

    if (!client_no_release_cid_flag)
        flags |= QMI_DEVICE_RELEASE_CLIENT_FLAGS_RELEASE_CID;
    else
        g_print ("[%s] Client ID not released:\n"
                 "\tService: '%s'\n"
                 "\t    CID: '%u'\n",
                 qmi_device_get_path_display (device),
                 qmi_service_get_string (service),
                 qmi_client_get_cid (client));

    qmi_device_release_client (device,
                               client,
                               flags,
                               10,
                               NULL,
                               (GAsyncReadyCallback)release_client_ready,
                               NULL);
}

static void
allocate_client_ready (QmiDevice *dev,
                       GAsyncResult *res)
{
    GError *error = NULL;

    client = qmi_device_allocate_client_finish (dev, res, &error);
    if (!client) {
        g_printerr ("error: couldn't create client for the '%s' service: %s\n",
                    qmi_service_get_string (service),
                    error->message);
        exit (EXIT_FAILURE);
    }

    /* Run the service-specific action */
    switch (service) {
    case QMI_SERVICE_DMS:
        qmicli_dms_run (dev, QMI_CLIENT_DMS (client), cancellable);
        return;
    case QMI_SERVICE_NAS:
        qmicli_nas_run (dev, QMI_CLIENT_NAS (client), cancellable);
        return;
    case QMI_SERVICE_WDS:
        qmicli_wds_run (dev, QMI_CLIENT_WDS (client), cancellable);
        return;
    case QMI_SERVICE_PBM:
        qmicli_pbm_run (dev, QMI_CLIENT_PBM (client), cancellable);
        return;
    case QMI_SERVICE_UIM:
        qmicli_uim_run (dev, QMI_CLIENT_UIM (client), cancellable);
        return;
    case QMI_SERVICE_WDA:
        qmicli_wda_run (dev, QMI_CLIENT_WDA (client), cancellable);
        return;
    default:
        g_assert_not_reached ();
    }
}

static void
device_allocate_client (QmiDevice *dev)
{
    guint8 cid = QMI_CID_NONE;

    if (client_cid_str) {
        guint32 cid32;

        cid32 = atoi (client_cid_str);
        if (!cid32 || cid32 > G_MAXUINT8) {
            g_printerr ("error: invalid CID given '%s'\n",
                        client_cid_str);
            exit (EXIT_FAILURE);
        }

        cid = (guint8)cid32;
        g_debug ("Reusing CID '%u'", cid);
    }

    /* As soon as we get the QmiDevice, create a client for the requested
     * service */
    qmi_device_allocate_client (dev,
                                service,
                                cid,
                                10,
                                cancellable,
                                (GAsyncReadyCallback)allocate_client_ready,
                                NULL);
}

static void
set_instance_id_ready (QmiDevice *dev,
                       GAsyncResult *res)
{
    GError *error = NULL;
    guint16 link_id;

    if (!qmi_device_set_instance_id_finish (dev, res, &link_id, &error)) {
        g_printerr ("error: couldn't set instance ID: %s\n",
                    error->message);
        exit (EXIT_FAILURE);
    }

    g_print ("[%s] Instance ID set:\n"
             "\tLink ID: '%" G_GUINT16_FORMAT "'\n",
             qmi_device_get_path_display (dev),
             link_id);

    /* We're done now */
    qmicli_async_operation_done (TRUE);
}

static void
device_set_instance_id (QmiDevice *dev)
{
    gint instance_id;

    if (g_str_equal (device_set_instance_id_str, "0"))
        instance_id = 0;
    else {
        instance_id = atoi (device_set_instance_id_str);
        if (instance_id == 0) {
            g_printerr ("error: invalid instance ID given: '%s'\n", device_set_instance_id_str);
            exit (EXIT_FAILURE);
        } else if (instance_id < 0 || instance_id > G_MAXUINT8) {
            g_printerr ("error: given instance ID is out of range [0,%u]: '%s'\n",
                        G_MAXUINT8,
                        device_set_instance_id_str);
            exit (EXIT_FAILURE);
        }
    }

    g_debug ("Setting instance ID '%d'...", instance_id);
    qmi_device_set_instance_id (dev,
                                (guint8)instance_id,
                                10,
                                cancellable,
                                (GAsyncReadyCallback)set_instance_id_ready,
                                NULL);
}

static void
get_service_version_info_ready (QmiDevice *dev,
                                GAsyncResult *res)
{
    GError *error = NULL;
    GArray *services;
    guint i;

    services = qmi_device_get_service_version_info_finish (dev, res, &error);
    if (!services) {
        g_printerr ("error: couldn't get service version info: %s\n",
                    error->message);
        exit (EXIT_FAILURE);
    }

    g_print ("[%s] Supported versions:\n",
             qmi_device_get_path_display (dev));
    for (i = 0; i < services->len; i++) {
        QmiDeviceServiceVersionInfo *info;
        const gchar *service_str;

        info = &g_array_index (services, QmiDeviceServiceVersionInfo, i);
        service_str = qmi_service_get_string (info->service);
        if (service_str)
            g_print ("\t%s (%u.%u)\n",
                     service_str,
                     info->major_version,
                     info->minor_version);
        else
            g_print ("\tunknown [0x%02x] (%u.%u)\n",
                     info->service,
                     info->major_version,
                     info->minor_version);
    }
    g_array_unref (services);

    /* We're done now */
    qmicli_async_operation_done (TRUE);
}

static void
device_get_service_version_info (QmiDevice *dev)
{
    g_debug ("Getting service version info...");
    qmi_device_get_service_version_info (dev,
                                         10,
                                         cancellable,
                                         (GAsyncReadyCallback)get_service_version_info_ready,
                                         NULL);
}

static void
device_open_ready (QmiDevice *dev,
                   GAsyncResult *res)
{
    GError *error = NULL;

    if (!qmi_device_open_finish (dev, res, &error)) {
        g_printerr ("error: couldn't open the QmiDevice: %s\n",
                    error->message);
        exit (EXIT_FAILURE);
    }

    g_debug ("QMI Device at '%s' ready",
             qmi_device_get_path_display (dev));

    if (device_set_instance_id_str)
        device_set_instance_id (dev);
    else if (get_service_version_info_flag)
        device_get_service_version_info (dev);
    else
        device_allocate_client (dev);
}

static void
device_new_ready (GObject *unused,
                  GAsyncResult *res)
{
    QmiDeviceOpenFlags open_flags = QMI_DEVICE_OPEN_FLAGS_NONE;
    GError *error = NULL;

    device = qmi_device_new_finish (res, &error);
    if (!device) {
        g_printerr ("error: couldn't create QmiDevice: %s\n",
                    error->message);
        exit (EXIT_FAILURE);
    }

    /* Setup device open flags */
    if (device_open_version_info_flag)
        open_flags |= QMI_DEVICE_OPEN_FLAGS_VERSION_INFO;
    if (device_open_sync_flag)
        open_flags |= QMI_DEVICE_OPEN_FLAGS_SYNC;
    if (device_open_proxy_flag)
        open_flags |= QMI_DEVICE_OPEN_FLAGS_PROXY;
    if (device_open_mbim_flag)
        open_flags |= QMI_DEVICE_OPEN_FLAGS_MBIM;
    if (device_open_net_str)
        if (!qmicli_read_net_open_flags_from_string (device_open_net_str, &open_flags))
            exit (EXIT_FAILURE);

    /* Open the device */
    qmi_device_open (device,
                     open_flags,
                     15,
                     cancellable,
                     (GAsyncReadyCallback)device_open_ready,
                     NULL);
}

/*****************************************************************************/

static void
parse_actions (void)
{
    guint actions_enabled = 0;

    /* Generic options? */
    if (generic_options_enabled ()) {
        service = QMI_SERVICE_CTL;
        actions_enabled++;
    }

    /* DMS options? */
    if (qmicli_dms_options_enabled ()) {
        service = QMI_SERVICE_DMS;
        actions_enabled++;
    }

    /* NAS options? */
    if (qmicli_nas_options_enabled ()) {
        service = QMI_SERVICE_NAS;
        actions_enabled++;
    }

    /* WDS options? */
    if (qmicli_wds_options_enabled ()) {
        service = QMI_SERVICE_WDS;
        actions_enabled++;
    }

    /* PBM options? */
    if (qmicli_pbm_options_enabled ()) {
        service = QMI_SERVICE_PBM;
        actions_enabled++;
    }

    /* UIM options? */
    if (qmicli_uim_options_enabled ()) {
        service = QMI_SERVICE_UIM;
        actions_enabled++;
    }

    /* WDA options? */
    if (qmicli_wda_options_enabled ()) {
        service = QMI_SERVICE_WDA;
        actions_enabled++;
    }

    /* Cannot mix actions from different services */
    if (actions_enabled > 1) {
        g_printerr ("error: cannot execute multiple actions of different services\n");
        exit (EXIT_FAILURE);
    }

    /* No options? */
    if (actions_enabled == 0) {
        g_printerr ("error: no actions specified\n");
        exit (EXIT_FAILURE);
    }

    /* Go on! */
}

int main (int argc, char **argv)
{
    GError *error = NULL;
    GFile *file;
    GOptionContext *context;

    setlocale (LC_ALL, "");

    g_type_init ();

    /* Setup option context, process it and destroy it */
    context = g_option_context_new ("- Control QMI devices");
    g_option_context_add_group (context,
                                qmicli_dms_get_option_group ());
    g_option_context_add_group (context,
                                qmicli_nas_get_option_group ());
    g_option_context_add_group (context,
                                qmicli_wds_get_option_group ());
    g_option_context_add_group (context,
                                qmicli_pbm_get_option_group ());
    g_option_context_add_group (context,
                                qmicli_uim_get_option_group ());
    g_option_context_add_group (context,
                                qmicli_wda_get_option_group ());
    g_option_context_add_main_entries (context, main_entries, NULL);
    if (!g_option_context_parse (context, &argc, &argv, &error)) {
        g_printerr ("error: %s\n",
                    error->message);
        exit (EXIT_FAILURE);
    }
    g_option_context_free (context);

    if (version_flag)
        print_version_and_exit ();

    g_log_set_handler (NULL, G_LOG_LEVEL_MASK, log_handler, NULL);
    g_log_set_handler ("Qmi", G_LOG_LEVEL_MASK, log_handler, NULL);
    if (verbose_flag)
        qmi_utils_set_traces_enabled (TRUE);

    /* No device path given? */
    if (!device_str) {
        g_printerr ("error: no device path specified\n");
        exit (EXIT_FAILURE);
    }

    /* Build new GFile from the commandline arg */
    file = g_file_new_for_commandline_arg (device_str);

    /* Setup signals */
    signal (SIGINT, signals_handler);
    signal (SIGHUP, signals_handler);
    signal (SIGTERM, signals_handler);

    parse_actions ();

    /* Create requirements for async options */
    cancellable = g_cancellable_new ();
    loop = g_main_loop_new (NULL, FALSE);

    /* Launch QmiDevice creation */
    qmi_device_new (file,
                    cancellable,
                    (GAsyncReadyCallback)device_new_ready,
                    GUINT_TO_POINTER (service));
    g_main_loop_run (loop);

    if (cancellable)
        g_object_unref (cancellable);
    if (client)
        g_object_unref (client);
    if (device)
        g_object_unref (device);
    g_main_loop_unref (loop);
    g_object_unref (file);

    return (operation_status ? EXIT_SUCCESS : EXIT_FAILURE);
}
